libs/gltfio/include/gltfio/SimpleViewer.h

```diff
@@ -0,0 +1,322 @@
+ * is header-only to avoid adding a filagui dependency to gltfio. Clients should use AssetLoader and
+ * ResourceLoader to load an asset before passing it into the setAsset() method.
+ */
+class SimpleViewer {
+public:
+    /**
+     * Constructs a SimpleViewer that has a fixed association with the given Filament objects. Upon
+     * construction, the simple viewer may create some additional Filament objects (such as light
+     * sources) that it owns.
+     */
+    SimpleViewer(filament::Engine* engine, filament::Scene* scene, filament::View* view);
+
+    /**
+     * Destroys the SimpleViewer and any Filament entities that it owns.
+     */
+    ~SimpleViewer();
+
+    /**
+     * Sets or changes the asset that is being viewed. The viewer automatically adds all the
+     * asset's entities into the scene. The viewer does not claim ownership over the asset or
+     * its entities.
+     */
+    void setAsset(FilamentAsset* asset, utils::NameComponentManager* names);
+
+    /**
+     * Sets or changes the current scene's IBL to allow the UI manipulate it.
+     * NOTE: this could be removed if we add a getter method to Scene.
+     */
+    void setIndirectLight(filament::IndirectLight* ibl);
+
+    /**
+     * Applies the currently-selected glTF animation to the transformation hierarchy and updates
+     * the bone matrices on all renderables.
+     */
+    void applyAnimation(double currentTime);
+
+    /**
+     * Constructs ImGui controls for the current frame and responds to everything that the user has
+     * changed since the previous frame.
+     *
+     * If desired this can be used in conjunction with the filagui library, which allows clients to
+     * render ImGui controls with Filament.
+     */
+    void updateUserInterface();
+
+    /**
+     * Retrieves the current width of the ImGui "window" which we are using as a sidebar.
+     * Clients can monitor this value to adjust the size of the view.
+     */
+    int getSidebarWidth() const { return mSidebarWidth; }
+
+    static constexpr int INITIAL_SIDEBAR_WIDTH = 350;
+
+private:
+    // Immutable properties set from the constructor.
+    filament::Engine* const mEngine;
+    filament::Scene* const mScene;
+    filament::View* const mView;
+    const utils::Entity mSunlight;
+
+    // Properties that can be changed from the application.
+    FilamentAsset* mAsset = nullptr;
+    utils::NameComponentManager* mNames = nullptr;
+    Animator* mAnimator = nullptr;
+    filament::IndirectLight* mIndirectLight = nullptr;
+
+    // Properties that can be changed from the UI.
+    int mCurrentAnimation = 1;
+    bool mResetAnimation = true;
+    float mIblIntensity = 30000.0f;
+    float mIblRotation = 0.0f;
+    bool mShowWireframe = false;
+    bool mEnableSunlight = true;
+    bool mEnableDithering = true;
+    int mSidebarWidth = INITIAL_SIDEBAR_WIDTH;
+};
+
+filament::math::mat4f fitIntoUnitCube(const filament::Aabb& bounds) {
```

Comment:
Why is the entire implementation in the header file?

Reply:
As mentioned in the class documentation, this is just header file because its use is optional and we don't want gltfio to have a hard dependency on imgui.

Reply:
Should we make this more like a header only library then, with #defines required to enable the implementation somewhere?

Reply:
Not a bad idea, I'll go ahead and add that in.

Reply:
Done.